codalab/lib/beam/streamingzipfile.py

```diff
@@ -0,0 +1,169 @@
+        None if there is no more available. This method is analogous to
+        TarFile.next().
+
+        We construct a ZipInfo object using the information stored in the next file header.
+        The logic here is based on the implementation of ZipFile._RealGetContents(), which
+        constructs a ZipInfo object from information in a central directory file header, but
+        modified to work with the file-header-specific struct
+        (for the implementation of ZipFile._RealGetContents(), see
+        https://github.com/python/cpython/blob/048f54dc75d51e8a1c5822ab7b2828295192aaa5/Lib/zipfile.py#L1316).
+        """
+        fp = self.fp
+
+        # First, advance to the next header, if needed.
+        fp.read(self._next_header_pos - fp.tell())
+
+        # Read the next header.
+        zipinfo = None
+        fheader = fp.read(sizeFileHeader)
+        if len(fheader) != sizeFileHeader:
+            raise BadZipFile("Truncated file header")
+        fheader = struct.unpack(structFileHeader, fheader)
+        if fheader[_FH_SIGNATURE] == stringCentralDir:
+            # We've reached the central directory. This means that we've finished iterating through
+            # all entries in the zip file. We can do this check because the file header signature
+            # and central directory signature are stored in the same spot (index 0) and with the same format.
+            self._loaded = True
+            return None
+        if fheader[_FH_SIGNATURE] != stringFileHeader:
+            raise BadZipFile("Bad magic number for file header")
+        filename = fp.read(fheader[_FH_FILENAME_LENGTH])
+        flags = fheader[_FH_GENERAL_PURPOSE_FLAG_BITS]
+        if flags & 0x800:
+            # UTF-8 file names extension
+            filename = filename.decode('utf-8')
+        else:
+            # Historical ZIP filename encoding
+            filename = filename.decode('cp437')
+        # Create ZipInfo instance to store file information
+        x = ZipInfo(filename)
+        x.extra = fp.read(fheader[_FH_EXTRA_FIELD_LENGTH])
+        x.header_offset = self._next_header_pos
+
+        # The file header stores nearly all the same information needed for ZipInfo as what the
+        # central directory file header stores, except for a couple of missing fields.
+        # We just set them to 0 here.
+        x.comment = 0
+        x.create_version, x.create_system = 0, 0
+        x.volume, x.internal_attr, x.external_attr = 0, 0, 0
+
+        (x.extract_version, x.reserved, x.flag_bits, x.compress_type, t, d,
+        x.CRC, x.compress_size, x.file_size) = fheader[1:10]
+        if x.extract_version > MAX_EXTRACT_VERSION:
+            raise NotImplementedError("zip file version %.1f" %
+                                    (x.extract_version / 10))
+
+        # Convert date/time code to (year, month, day, hour, min, sec)
+        x._raw_time = t
+        x.date_time = ( (d>>9)+1980, (d>>5)&0xF, d&0x1F,
```

Comment:
does it make sense to add a comment explaining this logic? simpletons like me may have a hard time understanding